### PR TITLE
Feature/mail notification on mapping changes

### PR DESCRIPTION
**Pull request by richard0000**
# What was done?
### Notify the involved users about changes in a mapped mapping
For finished mappings, a change now is informed to all involved users,
it, all those users that once worked on the mapping.

### Others
Using Action Mailer, implement the "forgot password" for any user.
Configure it using an account from environment variables.

# Screenshots
forgot-pass-working

image

# Related Issues
#60

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -44,12 +46,63 @@ def update_role(role_id)
...
+  def password_token_valid?
+    (reset_password_sent_at + 4.hours) > Time.now.utc
+  end
+
+  ###
+  # @description: Welcomes the user with an email which let's him sign in with
+  #   a new password.
+  ###
+  def send_welcome
+    UserMailer.with(user: User.last).welcome.deliver_now
```
Comment (excelsior) [inline app/models/user.rb:78]: Why does it send a welcome to the last user? This may result in a race condition.

File: app/models/user.rb [hunk 2]
```diff
@@ -44,12 +46,63 @@ def update_role(role_id)
...
+    super options.merge(methods: %i[roles organization])
+  end
+
+  private
+
+  ###
+  # @description: Generates a secure token
+  ###
+  def generate_token
+    SecureRandom.hex(10)
```
Comment (excelsior) [inline app/models/user.rb:106]: It's crucial to ensure that the generated token is unique.

File: config/environments/development.rb [hunk 3]
```diff
@@ -32,9 +32,23 @@
...
+  config.action_mailer.raise_delivery_errors = true
   config.action_mailer.perform_caching = false
+  config.action_mailer.delivery_method = :smtp
+  config.action_mailer.perform_deliveries = true
+  config.action_mailer.default_options = {
+    from: ENV["MAIL_USERNAME"],
+    host: ENV['API_URL'],
+    protocol: 'http'
+  }
+  config.action_mailer.smtp_settings = {
```
Comment (excelsior) [inline config/environments/development.rb:44]: I'd suggest using something like [letter_opener](https://github.com/ryanb/letter_opener) in development, if possible.

File: db/migrate/20201216115751_add_password_reset_columns_to_user.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class AddPasswordResetColumnsToUser < ActiveRecord::Migration[6.0]
+  def change
+    add_column :users, :reset_password_token, :string
```
Comment (excelsior) [inline db/migrate/20201216115751_add_password_reset_columns_to_user.rb:5]: The `reset_password_token` column must have a unique constraint.

**Review by excelsior — CHANGES_REQUESTED** [review]
Thanks, Ricardo. Looks mostly good 👍

A couple of questions:

1. This doesn't include Mailgun yet, it's coming later or?..
2. Will it be possible to use a layout template for HTML mailer views. Currently a lot of stuff is repeated in those views.

File: app/models/mapping.rb [hunk 5]
```diff
@@ -93,6 +93,28 @@ def update_selected_terms ids
...
+  def notify_updated
+    involved_users.each {|user|
+      MappingMailer.with(mapping: self, user: user).updated.deliver_now
+    }
+  end
+
+  ###
+  # @description: Get the users who worked in this mapping
+  ###
+  def involved_users
```
[the remark this answers is not in the document]
Comment (richard0000) [reply app/models/mapping.rb:108]: Love it

File: app/models/user.rb (hunk 2 above)
Comment (richard0000) [reply app/models/user.rb:106]: Will made a `tokenable` concern for that.

**Pull request merged**